libraries/AP_SurfaceDistance/AP_SurfaceDistance.cpp

```diff
@@ -0,0 +1,182 @@
+        glitch_count = 0;
+        alt_cm_glitch_protected = alt_cm;
+    }
+    if (abs(glitch_count) >= RANGEFINDER_GLITCH_NUM_SAMPLES) {
+        // clear glitch and record time so consumers (i.e. surface tracking) can reset their target altitudes
+        glitch_count = 0;
+        alt_cm_glitch_protected = alt_cm;
+        glitch_cleared_ms = now;
+        reset_terrain_offset = true;
+    }
+
+    // filter rangefinder altitude
+    const bool timed_out = now - last_healthy_ms > RANGEFINDER_TIMEOUT_MS;
+    if (alt_healthy) {
+        if (timed_out) {
+            // reset filter if we haven't used it within the last second
+            alt_cm_filt.reset(alt_cm);
+            reset_terrain_offset = true;
+            set_status(Status::Stale_Data);
+        } else {
+            alt_cm_filt.apply(alt_cm, 0.05f);
+        }
+        last_healthy_ms = now;
+    }
+
+    // handle reset of terrain offset
+    if (reset_terrain_offset) {
+        if (rotation == ROTATION_PITCH_90) {
+            // upward facing
+            terrain_offset_cm = inertial_alt_cm + alt_cm;
+        } else {
+            // assume downward facing
+            terrain_offset_cm = inertial_alt_cm - alt_cm;
+        }
+    }
+
+    Log_Write();
+}
+
+/*
+  get inertially interpolated rangefinder height. Inertial height is
+  recorded whenever we update the rangefinder height, then we use the
+  difference between the inertial height at that time and the current
+  inertial height to give us interpolation of height from rangefinder
+ */
+bool AP_SurfaceDistance::get_rangefinder_height_interpolated_cm(int32_t& ret) const
+{
+    if (!enabled_and_healthy()) {
+        return false;
+    }
+    ret = alt_cm_filt.get();
+    ret += inertial_nav.get_position_z_up_cm() - inertial_alt_cm;
+    return true;
+}
+
+void AP_SurfaceDistance::set_status(Status state_change)
+{
+    status |= uint8_t(state_change);
+}
+
+#if HAL_LOGGING_ENABLED
+void AP_SurfaceDistance::Log_Write(void) const
+{
+    // @LoggerMessage: SURF
+    // @Vehicles: Copter
+    // @Description: Surface distance measurement
+    // @Field: TimeUS: Time since system startup
+    // @Field: I: Instance
+    // @Field: St: Status
+    // @Field: D: Raw Distance
+    // @Field: FD: Filtered Distance
+    // @Field: TO: Terrain Offset
+
+    //Write to data flash log
+    AP::logger().WriteStreaming("SURF",
+                                "TimeUS,I,St,D,FD,TO",
+                                "s#-mmm",//units multipliers format
+                                "F--BBB",
+                                "QfBhhf",
+                                AP_HAL::micros64(),
+                                instance,
+                                status,
+                                alt_cm,
+                                alt_cm_filt.get(),
+                                terrain_offset_cm
+                                );
+}
+#endif  // HAL_LOGGING_ENABLED
+
+
+
+
+#endif // AP_RANGEFINDER_ENABLED
```

Comment:
Should be at the end?

Reply:
No, need to access the two helpers at the end of this file as copter does not define out the rangefinder state stuff currently (don't want to do that in this PR)